File: frontend/src/app/shared/dose.ts
export const DOSE_UNITS = [
  'mg',
  'mcg',
  'g',
  'mL',
  'IU',
  'units',
  'tablet',
  'capsule',
  'drop',
  'puff',
] as const;

export type DoseUnit = (typeof DOSE_UNITS)[number];

const DOSE_RE = /^\s*(\d+(?:\.\d+)?)\s*([A-Za-zµμ]+)?\s*$/;

function trimAmount(value: number): string {
  if (!Number.isFinite(value)) {
    return '';
  }
  return String(Number(value.toPrecision(12)));
}

export function parseDose(value: string | null | undefined): { amount: string | null; unit: string } {
  const text = (value ?? '').trim();
  if (!text) {
    return { amount: null, unit: 'mg' };
  }
  const match = DOSE_RE.exec(text);
  if (!match) {
    return { amount: null, unit: 'mg' };
  }
  return { amount: trimAmount(Number(match[1])), unit: match[2] || 'mg' };
}

export function composeDose(amount: string | number | null | undefined, unit: string | null | undefined): string {
  const text = String(amount ?? '').trim();
  if (!text) {
    return '';
  }
  const numeric = Number(text);
  const trimmed = Number.isFinite(numeric) ? trimAmount(numeric) : text;
  const suffix = (unit ?? 'mg').trim() || 'mg';
  return `${trimmed}${suffix}`;
}

export function administeredDose(unitDose: string | null | undefined, quantity: number | string | null | undefined): string {
  const parsed = parseDose(unitDose);
  if (parsed.amount === null) {
    return (unitDose ?? '').trim();
  }
  const qty = Number(quantity === null || quantity === undefined || quantity === '' ? 1 : quantity);
  const safeQty = Number.isFinite(qty) && qty > 0 ? qty : 1;
  return composeDose(Number(parsed.amount) * safeQty, parsed.unit);
}

File: backend/app/services/dose.py
from __future__ import annotations

import re
from decimal import Decimal, InvalidOperation

DOSE_UNITS = (
    "mg",
    "mcg",
    "g",
    "mL",
    "IU",
    "units",
    "tablet",
    "capsule",
    "drop",
    "puff",
)

_DOSE_RE = re.compile(r"^\s*(\d+(?:\.\d+)?)\s*([A-Za-zµμ]+)?\s*$")


def parse_dose(value: str | None) -> tuple[str | None, str]:
    text = (value or "").strip()
    if not text:
        return None, "mg"
    match = _DOSE_RE.match(text)
    if not match:
        return None, "mg"
    return _trim(Decimal(match.group(1))), match.group(2) or "mg"


def compose_dose(amount: object, unit: str | None) -> str:
    text = str(amount or "").strip()
    if not text:
        return ""
    try:
        trimmed = _trim(Decimal(text))
    except InvalidOperation:
        trimmed = text
    suffix = (unit or "mg").strip() or "mg"
    return f"{trimmed}{suffix}"


def administered_dose(unit_dose: str | None, quantity: object | None) -> str:
    amount, unit = parse_dose(unit_dose)
    if amount is None:
        return str(unit_dose or "").strip()
    try:
        qty = Decimal(str(quantity if quantity not in (None, "") else 1))
    except InvalidOperation:
        qty = Decimal(1)
    if qty <= 0:
        qty = Decimal(1)
    return compose_dose(Decimal(amount) * qty, unit)


def _trim(value: Decimal) -> str:
    text = format(value.normalize(), "f")
    if "." in text:
        text = text.rstrip("0").rstrip(".")
    return text or "0"
